CODING_RULES.md

```diff
@@ -11,90 +11,156 @@ to be completed
 
 - use getters/setters instead of `@property`
 
-### regarding the exceptions
-
-
-- use exceptions defined in **fedbiomed.common.exceptions**
-
-- callee side: then detecting a python exception in a fedbiomed layer :
-
-  - print a logger.*() message
-
-    - could be logger.critical -> stop the software, cannot continue
-      (ex: FedbiomedEnvironError)
-    - or logger.error -> software may continue (ex: MessageError)
-
-  - raise the exception as a fedbiomed exception
-
-
-- caller/intermediate side: trap the exceptions:
-
-  - separate the FedbiomedError and other exceptions
-
-```
-try:
-
-    something()
-
-except SysError as e:
-    logger.XXX()
-    raise OneOfFedbiomedError()
-```
-
-  - example of top level function (eg: Experiment())
-
-```
-try:
-
-    something()
-
-except FedbiomedError as e:
-    etc...
-
-except Exception as e:   <=== objective: minimize the number of type we arrive here !
-    # place to do a backtrace
-    # extra message to the end user to post this backtrace to the support team
-    etc...
-```
-
-  - except of the top level program, it is **forbidden** to trap all exceptions (with ```except:``` or ```except Exception```)
-
-
-- the **try:** block is as small as possible
-
-- force to read the documentation
-
+### exceptions handling
+
+- on the node: in general, node should not stop because of exceptions that occur during its operation
+
+- on the researcher: general behaviour is to propagate the exceptions to the top level layer, where they are transformed to a friendlier output. Researcher displays this output and stops.
+
+- when a class raises an exception, it raises a FedbiomedSomethingError: use exceptions defined in **fedbiomed.common.exceptions** :
+
+  Do:
+  ```
+  raise FedbiomedSometypeError()
+  ```
+
+  Don't:
+  ```
+  raise NameError()
+  ```
+
+- optionally, if more specificity is wanted, a class can catch a python (non-Fed-Biomed) exception and re-raise it as a FedbiomedError
+
+  Optionally do:
+  ```
+  try:
+      something()
+  except SysError as e:
+      raise OneOfFedbiomedError()
+  ```
+
+- a class generally shouldn't catch a FedbiomedError and re-raise (a FedbiomedError):
+
+  Don't:
+  ```
+  try:
+    somecode()
+  except FedbiomedSometypeError;
+    raise FedbiomedOnetypeError()
+  ```
+
+- when catching exceptions
+
+  - try to be specific about the exception if easy/meaningful:
+
+    Ideally:
+    ```
+    try:
+      mycode()
+    except ValueError as e:
+      ...
+    ```
+
+    If needed:
+    ```
+    try:
+      mycode()
+    except Exception as e:
+      ...
+    ```
+
+  - don't use the `except:` clause
+
+    Don't
+    ```
+    try:
+      mycode()
+    except:
+      ...
+    ```
+
+  - should separate FedbiomedError and other exceptions
+
+    Do:
+    ```
+    try:
+      mycode()
+    except FedbiomedSomeError as e:
+      ...
+    except FedbiomedError as e:
+      ...
+    except Exception as e:
```

Comment:
Hi @mvesin, 

We can also specify that if all other exceptions different than `FedbiomedError` are considered as undetermined error and needs to be cached at the top layer where the first request is executed. For example, `Node.task_manager.train_task`

Reply:
Thank you @srcansiz 
Can you please reformulate, I don't understand your comment ? (sentence seems unfinished)

Reply:
What I meant is that `except Exception` cases are undetermined error cases that are usually not expected. It can be due to anything. Therefore,  we should consider those as `Undefined` error while sending back to researcher without attaching the exception message.  These exceptions should always be caught where the task is executed (top-layer: where `round.run_model_training` or `secagg.setup`  methods are executed) rather than  lower layers. This is of course not the case if developer wants to add some specificity to the undefined error message from the lower layers.

Reply:
Thank you @srcansiz 

Pushed a new version that (hopefully) addresses these points
- `except Exception:` in top layer
- `except Exception:` in lower layers
- node sending errors to researcher in the case of exception caught at top layer